### ISSUE TITLE
Including Release Notes in sympy repo.

### ISSUE THREAD

**Issue by leosartaj** [sympy/sympy#10942]
Is there a specific reason why we don't have release notes version controlled? If version controlled, we can then also include them in the docs which seems missing to me.

**Comment by leosartaj** [comment]
Ping @asmeurer @certik

**Comment by certik** [comment]
I think that would be a good idea. I don't see why not.

**Comment by leosartaj** [comment]
> I asked about this on the mailing list or github years ago. Aaron was against it because you constantly have merge conflicts, which is true. But scipy and such have them in the repo. 

 I recently noticed this with the git project. AFAIK almost all major projects inlcude the notes in the repo.

> They may have one file per note and then merge them all for the docs build so that you won't have merge conflicts when multiple people edit one release notes file.

This could actually work. Probably store all the files in a directory which are ultimately merged to build docs.

**Comment by asmeurer** [comment]
I'm for whatever gets people to keep the release notes up to date.

**Comment by moorepants** [comment]
In PyDy we have an issue template that has a checkbox for updating the
release note. It has added a surprising amount of accountability and pretty
much everything has been added.


Jason
moorepants.info
+01 530-601-9791

**Comment by asmeurer** [comment]
So here's an update on this issue.  

Presently, the release notes are added to wiki pages at https://github.com/sympy/sympy/wiki/Release-Notes. 

The primary goal here is

1. Every pull request should update the release notes with major changes that it makes.

There is also a secondary goal, which is:

2. Put the release notes in the SymPy docs (instead of the wiki)

I am personally less convinced that this second goal is necessary, but if it falls out nicely from the process, we can do it. 

The primary issue for 1 is that there is no mechanism for preventing PRs from being merged without a release notes entry being added (no tests fail), so very few people do it. There are also no reminders so people generally don't even do it voluntarily. 

These are the suggestions so far for doing the release notes process. 

A0 (current process). Release notes are added to a respective wiki page at https://github.com/sympy/sympy/wiki/Release-Notes. 

**Benefits**

- Anyone can edit the wiki. 
- No pull requests required to edit.
- Old release notes can be edited as well.

**Downsides**

- No current checks that the release notes are edited, so very few do so.
- Release notes are divorced from the docs. 

A. (from https://groups.google.com/forum/#!msg/sympy/6eJW2qNT6dk/abmqUYP0AAAJ). Every pull request adds a new file to a changelog directory, based on a template, with the release notes for that PR. At release time (or maybe also at docs build time), the changelog files are amalgamated into a single file. The single file part is important, as it avoids git merge conflicts. 

This is the same process that xonsh uses, so we could probably borrow some of their tooling, including a PR linter. 

**Benefits**
- Release notes are in the main docs
- Existing tooling
- Avoids git conflicts

**Downsides**
- Complicated process, especially for new contributors
- Release notes must be amalgamated before they can be human consumed. 
- Even though xonsh tooling exists we still need to adapt it to SymPy.

B. (suggested by @ylemkimon at https://github.com/sympy/sympy/pull/13560). We add something like

```
---
# Changes
- Please list major changes for this PR here. If there are no major changes, add "NO MAJOR CHANGES".
```

to the PR template. We then have some tooling which checks that this template is filled out, say in the Travis tests. 

**Benefits**
- Extremely simple process, especially for new contributors
- Release notes could be amalgamated in the docs or the wiki

**Downsides**
- Release notes must still be amalgamated.
- We would need to write tooling around it (checks, amalgamation)
- Release notes would necessarily be written in Markdown, so amalgamation into the RST docs would require extra work.

C. (suggested by me at https://github.com/sympy/sympy/pull/13560). Basically, we keep the current process (editing on the wiki), but we also add some Travis checks that each PR adds an entry. We could also potentially move the release notes over to the docs at release time. 

**Benefits**
- All the same benefits of A0 (easy to edit, no PR required, old release notes can be edited)
- Minimal tooling would need to be written. 
- No changes required for existing release notes

**Downsides**
- Compared to suggestion B, it's a little less easy for new contributors, and also less easy for reviewers to check.
- Tooling required.
- Unclear how to test that release notes have been added (just check for link back to PR?). Reviewer needs to check to see it the added release notes are good. How to allow no release notes added, for trivial PRs?

---

I'm personally leaning toward option C because it seems the easiest, but if someone is willing to do the work to implement one of the other options I am not opposed to any of them. Or if someone has another suggestion I'm open to hearing it.

**Comment by moorepants** [comment]
The benefit for having them on the wiki is stated as "anyone can edit them". I don't see this as a benefit. My guess is that we have had few to no people that haven't submitted a PR editing the release notes on the wiki. The wiki is a big mess of information that is hard to navigate and I've yet to find the release notes from another project that are not in the source repository. Having release notes in the repository is essential in my opinion. It is the only way to determine differences in versions if you have only the source tarballs. I think we should go with A and just adopt what another project is doing. All of the main scipy stack projects have some form of getting the release notes into the repo. We can copy any number of them. Lastly, the real issue is how to build a culture (or requirement) of superb release notes. Projects like Pandas, matplotlib, and Scipy are really blowing us away on this.  I think the merge conflict issue is not as big as it is made out to be, btw.

tldr,

1. Adopt any major scipy stack project's method of adding release notes to the source repo, suck it up and deal with any warts the method has.
2. Institute a PR merge requirement that the release notes are updated before merge (can be automated or reviewers should be instructed to require this before merging)
3. Brainstorm ways to get our release notes up to higher quality like our peers.

**Comment by asmeurer** [comment]
An example of why "anyone can edit them" is a benefit https://github.com/sympy/sympy/issues/13385.

**Comment by ylemkimon** [comment]
@asmeurer It seem that only those with push access to the repository can push to the wiki, even if it is publicly editable. Will it be dangerous to include access token or deploy key that has push access to sympy/sympy in Travis?

**Comment by asmeurer** [comment]
It is possible to add a deploy key to Travis so that it's encrypted (use the `travis encrypt` command). It will only be available to branches on sympy/sympy (not pull requests). So that mitigates the danger because malicious code would have to be merged first. 

I don't know if deploy keys give you push access to the wiki. I'm assuming they do. 

We could also potentially use [doctr](https://drdoctr.github.io/) to manage the deploy keys, though I would need to update it first to support wikis.

**Comment by ylemkimon** [comment]
Is there are a Sphinx extension to retrieve external resources and fallback if Internet connection is not available?

**Comment by asmeurer** [comment]
I'm not familiar with one but that doesn't mean it doesn't exist.
